narwhals/_expression_parsing.py

```diff
@@ -399,3 +357,105 @@ def evaluate_output_names_and_aliases(
             *[(x, alias) for x, alias in zip(output_names, aliases) if x not in exclude]
         )
     return output_names, aliases
+
+
+def operation_is_order_dependent(*args: IntoExpr | Any) -> bool:
+    from narwhals.expr import Expr
+
+    return any(isinstance(x, Expr) and x._metadata["is_order_dependent"] for x in args)
+
+
+class ExprKind(Enum):
+    """Describe which kind of expression we are dealing with.
+
+    Commutative composition rules are:
+    - LITERAL vs LITERAL -> LITERAL
+    - CHANGES_LENGTH vs (LITERAL | AGGREGATION) -> CHANGES_LENGTH
+    - CHANGES_LENGTH vs (CHANGES_LENGTH | TRANSFORM) -> raise
+    - TRANSFORM vs (LITERAL | AGGREGATION) -> TRANSFORM
+    - AGGREGATION vs (LITERAL | AGGREGATION) -> AGGREGATION
+    """
+
+    LITERAL = auto()  # e.g. nw.lit(1)
+    AGGREGATION = auto()  # e.g. nw.col('a').mean()
+    TRANSFORM = auto()  # length-preserving, e.g. nw.col('a').round()
+    CHANGES_LENGTH = auto()  # e.g. nw.col('a').drop_nulls()
+
+
+class ExprMetadata(TypedDict):
+    kind: ExprKind
+    is_order_dependent: bool
+
+
+def combine_metadata(*args: IntoExpr, strings_are_column_names: bool) -> ExprMetadata:
+    # Combine metadata from `args`.
+    from narwhals.expr import Expr
+
+    n_changes_length = 0
+    n_transforms = 0
+    n_aggregations = 0
+    n_literals = 0
+    is_order_dependent = False
+
+    for arg in args:
+        if isinstance(arg, str) and strings_are_column_names:
+            n_transforms += 1
+        elif isinstance(arg, Expr):
+            if arg._metadata["is_order_dependent"]:
+                is_order_dependent = True
```

Comment:
minor nit 😉

```suggestion
            is_order_dependent = arg._metadata["is_order_dependent"]
```

Liking the switch to `Enum` btw

Reply:
yeah, if any expression there is order-dependent, then the result expression is order-dependent